libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/route/line/api/MapboxRouteLineView.kt

```diff
@@ -80,120 +148,133 @@ class MapboxRouteLineView(var options: MapboxRouteLineOptions) {
-                                it.dynamicData.baseExpressionProvider
-                            )
-                            updateLineGradientAsync(
-                                jobControl.scope,
+                    val sourceFeaturePairings = sourceToFeatureMap.toMutableList()
+                    routeLineDatas.mapIndexed { index, routeLineData ->
+                        val relatedSourceKey = getRelatedSourceKey(
+                            routeLineData.featureCollection.features()?.firstOrNull()?.id(),
+                            sourceFeaturePairings
+                        ).also { sourceFeaturePairings.remove(it) }
+                        ifNonNull(relatedSourceKey) { sourceKeyFeaturePair ->
+                            val gradientCommands = getGradientUpdateCommands(
                                 style,
-                                RouteLayerConstants.ALTERNATIVE_ROUTE2_CASING_LAYER_ID,
-                                it.dynamicData.casingExpressionProvider
+                                sourceKeyFeaturePair.first,
+                                routeLineData,
+                                sourceLayerMap
                             )
-                            updateLineGradientAsync(
-                                jobControl.scope,
-                                style,
-                                RouteLayerConstants.ALTERNATIVE_ROUTE2_TRAFFIC_LAYER_ID,
-                                it.dynamicData.trafficExpressionProvider
-                            )
-                        }
+                            // Ignoring the trim offsets if the source was updated with an
+                            // empty feature collection. Even though the call to update the
+                            // source was made first the trim offset commands seem to get
+                            // rendered before the source update completes resulting in an
+                            // undesirable flash on the screen.
+                            val trimOffsetCommands = when (sourceKeyFeaturePair.second.id()) {
+                                null -> listOf()
+                                else -> getTrimOffsetCommands(
+                                    style,
+                                    sourceKeyFeaturePair.first,
+                                    routeLineData,
+                                    sourceLayerMap
+                                )
+                            }
+                            val layerMoveCommand = if (index == 0) {
+                                {
+                                    moveLayersUp(
+                                        style,
+                                        sourceKeyFeaturePair.first,
+                                        sourceLayerMap
+                                    )
+                                }
+                            } else { {} }
```

Comment:
The layer move command is only executed once, when the loop index is 0.

Reply:
> I am also noticing that the route case is updated first then the route body, can we update them the other way around?

I can change the ordering.

Reply:
> The layer move command is only executed once, when the loop index is 0.
That's weird 🤔 , when I am debugging, within the same `renderRouteDrawData`, this command is called three times.

Reply:
When I log the `moveStyleLayer` calls when switching between routes I see only a single layer group (with 6 layers) getting executed. So after drawing the routes and selecting the alternative route I see:

style.moveStyleLayer(mapbox-layergroup-2-trailCasing)
style.moveStyleLayer(mapbox-layergroup-2-trail)
style.moveStyleLayer(mapbox-layergroup-2-casing)
style.moveStyleLayer(mapbox-layergroup-2-main)
style.moveStyleLayer(mapbox-layergroup-2-traffic)
style.moveStyleLayer(mapbox-layergroup-2-restricted)

That's what I would expect.  Are you seeing something different?

Reply:
I just double-checked, yes the effective moving layers calls are only called once, sorry for the confusion.

By adding logs, I find that the line gradient update is not grouped as layer groups, something like the following snippet shows, would this affect the color update which finally leads to the issue Lukas mentioned?
```
2022-06-09 23:39:07.275 13559-13559/com.mapbox.navigation.qa_test_app V/testing: --------------------------
2022-06-09 23:39:07.275 13559-13559/com.mapbox.navigation.qa_test_app V/testing: mapbox-layerGroup-1-trailCasing
2022-06-09 23:39:07.275 13559-13559/com.mapbox.navigation.qa_test_app V/testing: mapbox-layerGroup-1-trail
2022-06-09 23:39:07.275 13559-13559/com.mapbox.navigation.qa_test_app V/testing: mapbox-layerGroup-1-casing
2022-06-09 23:39:07.276 13559-13559/com.mapbox.navigation.qa_test_app V/testing: mapbox-layerGroup-1-main
2022-06-09 23:39:07.276 13559-13559/com.mapbox.navigation.qa_test_app V/testing: mapbox-layerGroup-1-traffic
2022-06-09 23:39:07.276 13559-13559/com.mapbox.navigation.qa_test_app V/testing: mapbox-layerGroup-1-restricted
2022-06-09 23:39:07.300 13559-13559/com.mapbox.navigation.qa_test_app V/testing: Update line gradient for id: mapbox-layerGroup-1-restricted
2022-06-09 23:39:07.301 13559-13559/com.mapbox.navigation.qa_test_app V/testing: Update line gradient for id: mapbox-layerGroup-1-main
2022-06-09 23:39:07.302 13559-13559/com.mapbox.navigation.qa_test_app V/testing: Update line gradient for id: mapbox-layerGroup-1-trail
2022-06-09 23:39:07.304 13559-13559/com.mapbox.navigation.qa_test_app V/testing: Update line gradient for id: mapbox-layerGroup-1-casing
2022-06-09 23:39:07.306 13559-13559/com.mapbox.navigation.qa_test_app V/testing: Update line gradient for id: mapbox-layerGroup-2-restricted
2022-06-09 23:39:07.307 13559-13559/com.mapbox.navigation.qa_test_app V/testing: Update line gradient for id: mapbox-layerGroup-1-trailCasing
2022-06-09 23:39:07.308 13559-13559/com.mapbox.navigation.qa_test_app V/testing: Update line gradient for id: mapbox-layerGroup-2-main
2022-06-09 23:39:07.310 13559-13559/com.mapbox.navigation.qa_test_app V/testing: Update line gradient for id: mapbox-layerGroup-2-trail
2022-06-09 23:39:07.311 13559-13559/com.mapbox.navigation.qa_test_app V/testing: Update line gradient for id: mapbox-layerGroup-2-trailCasing
2022-06-09 23:39:07.312 13559-13559/com.mapbox.navigation.qa_test_app V/testing: Update line gradient for id: mapbox-layerGroup-3-restricted
2022-06-09 23:39:07.312 13559-13559/com.mapbox.navigation.qa_test_app V/testing: Update line gradient for id: mapbox-layerGroup-2-casing
2022-06-09 23:39:07.313 13559-13559/com.mapbox.navigation.qa_test_app V/testing: Update line gradient for id: mapbox-layerGroup-3-casing
2022-06-09 23:39:07.314 13559-13559/com.mapbox.navigation.qa_test_app V/testing: Update line gradient for id: mapbox-layerGroup-3-trail
2022-06-09 23:39:07.315 13559-13559/com.mapbox.navigation.qa_test_app V/testing: Update line gradient for id: mapbox-layerGroup-3-main
2022-06-09 23:39:07.317 13559-13559/com.mapbox.navigation.qa_test_app V/testing: Update line gradient for id: mapbox-layerGroup-3-trailCasing
2022-06-09 23:39:07.319 13559-13559/com.mapbox.navigation.qa_test_app V/testing: Update line gradient for id: mapbox-layerGroup-1-traffic
2022-06-09 23:39:07.321 13559-13559/com.mapbox.navigation.qa_test_app V/testing: Update line gradient for id: mapbox-layerGroup-2-traffic
2022-06-09 23:39:27.517 13559-13559/com.mapbox.navigation.qa_test_app V/testing: --------------------------
2022-06-09 23:39:27.517 13559-13559/com.mapbox.navigation.qa_test_app V/testing: mapbox-layerGroup-2-trailCasing
2022-06-09 23:39:27.518 13559-13559/com.mapbox.navigation.qa_test_app V/testing: mapbox-layerGroup-2-trail
2022-06-09 23:39:27.518 13559-13559/com.mapbox.navigation.qa_test_app V/testing: mapbox-layerGroup-2-casing
2022-06-09 23:39:27.518 13559-13559/com.mapbox.navigation.qa_test_app V/testing: mapbox-layerGroup-2-main
2022-06-09 23:39:27.519 13559-13559/com.mapbox.navigation.qa_test_app V/testing: mapbox-layerGroup-2-traffic
2022-06-09 23:39:27.519 13559-13559/com.mapbox.navigation.qa_test_app V/testing: mapbox-layerGroup-2-restricted
2022-06-09 23:39:27.536 13559-13559/com.mapbox.navigation.qa_test_app V/testing: Update line gradient for id: mapbox-layerGroup-2-restricted
2022-06-09 23:39:27.538 13559-13559/com.mapbox.navigation.qa_test_app V/testing: Update line gradient for id: mapbox-layerGroup-2-main
2022-06-09 23:39:27.540 13559-13559/com.mapbox.navigation.qa_test_app V/testing: Update line gradient for id: mapbox-layerGroup-2-casing
2022-06-09 23:39:27.542 13559-13559/com.mapbox.navigation.qa_test_app V/testing: Update line gradient for id: mapbox-layerGroup-2-trail
2022-06-09 23:39:27.543 13559-13559/com.mapbox.navigation.qa_test_app V/testing: Update line gradient for id: mapbox-layerGroup-2-trailCasing
2022-06-09 23:39:27.545 13559-13559/com.mapbox.navigation.qa_test_app V/testing: Update line gradient for id: mapbox-layerGroup-1-restricted
2022-06-09 23:39:27.552 13559-13559/com.mapbox.navigation.qa_test_app V/testing: Update line gradient for id: mapbox-layerGroup-2-traffic
2022-06-09 23:39:27.553 13559-13559/com.mapbox.navigation.qa_test_app V/testing: Update line gradient for id: mapbox-layerGroup-1-main
2022-06-09 23:39:27.554 13559-13559/com.mapbox.navigation.qa_test_app V/testing: Update line gradient for id: mapbox-layerGroup-1-casing
2022-06-09 23:39:27.555 13559-13559/com.mapbox.navigation.qa_test_app V/testing: Update line gradient for id: mapbox-layerGroup-1-trail
2022-06-09 23:39:27.556 13559-13559/com.mapbox.navigation.qa_test_app V/testing: Update line gradient for id: mapbox-layerGroup-1-trailCasing
2022-06-09 23:39:27.557 13559-13559/com.mapbox.navigation.qa_test_app V/testing: Update line gradient for id: mapbox-layerGroup-3-main
2022-06-09 23:39:27.558 13559-13559/com.mapbox.navigation.qa_test_app V/testing: Update line gradient for id: mapbox-layerGroup-3-casing
2022-06-09 23:39:27.559 13559-13559/com.mapbox.navigation.qa_test_app V/testing: Update line gradient for id: mapbox-layerGroup-3-trail
2022-06-09 23:39:27.560 13559-13559/com.mapbox.navigation.qa_test_app V/testing: Update line gradient for id: mapbox-layerGroup-3-trailCasing
2022-06-09 23:39:27.561 13559-13559/com.mapbox.navigation.qa_test_app V/testing: Update line gradient for id: mapbox-layerGroup-3-restricted
2022-06-09 23:39:27.563 13559-13559/com.mapbox.navigation.qa_test_app V/testing: Update line gradient for id: mapbox-layerGroup-1-traffic
```

Reply:
I wouldn't expect the gradient updates to occur in any predictable order.  The `Expression` for each gradient is calculated asynchronously and as each finishes a call is made to update the line layer gradient.  The traffic `Expression` is the most complex so it will be the slowest.

One thing I just remembered is the maps call to update the line gradient is async as well.   The view class that's making all of these gradient updates and layer movements is first executing the gradient update commands, which are async. then executing the move layer commands. I don't know if the move layer commands are async or synchronous in the maps API.  But maybe what is seen in the map is some gradient updates, then some layer movements up the stack, then more gradient updates, then move layer movements up the stack etc. since all those calls aren't synchronous with in the maps code.  I'll make an adjustment to move the layers first before executing the  gradient updates to see if that makes any difference.

Reply:
As I can't recreate the issue demonstrated in @LukasPaczos 's video perhaps both of you could take a look with your devices.

Reply:
I am still seeing the same problem with the latest update:

https://user-images.githubusercontent.com/11026055/173029545-7c0ca347-bcf6-45a7-83fa-7aabb7e730f0.mp4

Reply:
With a short route, there seems to be no above problem. But for the longer route, this problem exists. As you mentioned above, the gradient calculation is async, and for the long route, it may take an even longer time for the calculation. As you can see from the log above, the time difference between the last move layer cmd to the last line gradient update cmd is about ~50ms, which will make the route update end up with different frames.

Reply:
Are you saying all of the layer move calls, the trim-offset calls and the gradient update calls have to be made within some specific time frame? The delay seen in the video is more than 50ms.

Reply:
I replaced the lazy `Expression` calculation with eager calculation and wait for all of them to complete before making any map update calls.  Now all the map update calls including layer moves, trim-offsets, and gradient updates happen sequentially in 6-15ms on my device. I'm not sure what else I can do.

Reply:
> I replaced the lazy Expression calculation with eager calculation and wait for all of them to complete before making any map update calls. Now all the map update calls including layer moves, trim-offsets, and gradient updates happen sequentially in 6-15ms on my device. I'm not sure what else I can do.

This approach works very well as with the latest commit I do not see any update disalignment as more 🎉